core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java

```diff
@@ -560,6 +592,252 @@ public UpdateNode plan(Update node)
                 outputs);
     }
 
+    private MergeCaseKind getMergeCaseKind(MergeCase mergeCase)
+    {
+        requireNonNull(mergeCase, "mergeCase is null");
+        if (mergeCase instanceof MergeInsert) {
+            return MergeCaseKind.INSERT;
+        }
+        if (mergeCase instanceof MergeUpdate) {
+            return MergeCaseKind.UPDATE;
+        }
+        if (mergeCase instanceof MergeDelete) {
+            return MergeCaseKind.DELETE;
+        }
+        throw new IllegalArgumentException("Unrecognized MergeCase " + mergeCase.getClass());
+    }
+
+    public MergeNode plan(Merge merge)
+    {
+        MergeAnalysis mergeAnalysis = analysis.getMergeAnalysis().orElseThrow(() -> new IllegalArgumentException("analysis.getMergeAnalysis() isn't present"));
+
+        List<ColumnHandle> redistributionColumns = mergeAnalysis.getRedistributionColumnHandles();
+        Map<Integer, List<ColumnHandle>> mergeCaseColumnsHandles = mergeAnalysis.getMergeCaseColumnHandles();
+
+        MergeDetails mergeDetails = createMergeDetails(merge, mergeCaseColumnsHandles);
+
+        RelationPlan target = new RelationPlanner(analysis, symbolAllocator, idAllocator, lambdaDeclarationToSymbolMap, metadata, outerContext, session, recursiveSubqueries)
+                .process(merge.getTable());
+
+        PlanBuilder targetPlanBuilder = newPlanBuilder(target, analysis, lambdaDeclarationToSymbolMap);
+
+        Assignments.Builder projections = Assignments.builder();
+        projections.putIdentities(targetPlanBuilder.getRoot().getOutputSymbols());
+
+        Symbol presentColumn = symbolAllocator.newSymbol("present", BOOLEAN);
+        projections.put(presentColumn, TRUE_LITERAL);
+
+        target = new RelationPlan(
+                new ProjectNode(idAllocator.getNextId(), targetPlanBuilder.getRoot(), projections.build()),
+                analysis.getScope(merge.getTable()),
```

Comment:
I **think** you mean that targetTableScope from `StatementAnalyzer` should be preserved in `MergeAnalysis` and used here.  Done.

Reply:
Actually, that was not my intention.
The `StatementAnalyzer` is supposed to assign scopes to nodes, so it should be possible to get the scope directly from the Analysis, like you did before.
The issue was in that the scope recorded in the Analysis for the `merge.getTable()` node is potentially not the right scope. It doesn't regard the table alias.

The solution to this would be to change the `Merge` node so that instead of two fields:
```
    private final Table table;
    private final Optional<Identifier> targetAlias;
```
it has one field:
```
    private final Relation target;
```
That field could be a `Table`, or an `AliasedRelation` in case when there was an alias. This is similar to the field `source`.

Then if you did:
```
Scope targetTableScope = analyzer.analyzeForUpdate(target, scope, UpdateKind.MERGE);
```
you would get the right scope and you would no more have to "amend" it by applying the alias.

Reply:
I believe I've done this in the latest commit.